### PR TITLE
GEODE-2578: Remove 64 KiB limit on query strings.

### PR DESCRIPTION
**Pull request by PivotalSarge**
- Remove artificial cap of 65535 for query string
  length by using 32 bits for the length of query
  strings in DataOutput::writeFullUTF().
- Rename parameter to TcrMessage::writeStringPart()
  whose name is misleading due to copy-and-paste.
- Add three unit tests for query-related messages:
  * testConstructorEXECUTECQ_MSG_TYPE
  * testConstructorWithGinormousQueryEXECUTECQ_MSG_TYPE
  * testConstructorEXECUTECQ_WITH_IR_MSG_TYPE
- Fix the cursor advance tests to not depend on
  specific values for buffer length and not to
  test remaining buffer length.

### CODE REVIEW

File: src/cppcache/include/geode/DataOutput.hpp [hunk 1]
```diff
@@ -378,9 +378,8 @@ class CPPCACHE_EXPORT DataOutput {
      */
   inline void writeFullUTF(const char* value, uint32_t length = 0) {
     if (value != NULL) {
-      int32_t len = getEncodedLength(value, length);
-      uint16_t encodedLen = static_cast<uint16_t>(len > 0xFFFF ? 0xFFFF : len);
```
Comment (echobravopapa) [inline src/cppcache/include/geode/DataOutput.hpp]: what was the corner case impact of this line?

File: src/cppcache/test/DataOutputTest.cpp [hunk 2]
```diff
@@ -305,15 +305,9 @@ TEST_F(DataOutputTest, TestCursorAdvance) {
       "001B596F7520686164206D65206174206D65617420746F726E61646F2E",
       dataOutput.getByteArray());
 
-  EXPECT_EQ((2 + 27), dataOutput.getBufferLength());
-
-  // buffers are pre-allocated 8k and have 2 bytes to hold the data length
-  EXPECT_EQ(((8 * 1024) - (2 + 27)), dataOutput.getRemainingBufferLength());
```
Comment (echobravopapa) [inline src/cppcache/test/DataOutputTest.cpp]: What led you to removing the call to getRemainingBufferLength()?

File: src/cppcache/test/TcrMessage_unittest.cpp [hunk 3]
```diff
@@ -559,3 +560,66 @@ TEST_F(TcrMessageTest, testConstructorEXECUTE_FUNCTION) {
...
+  EXPECT_EQ(TcrMessage::EXECUTECQ_MSG_TYPE, testMessage.getMessageType());
+
+  EXPECT_MESSAGE_EQ(
+      "0000002A0000004000000005FFFFFFFF0000000009004578656375746543510000001800"
+      "73656C656374202A2066726F6D202F736F6D657768657265000000040000000001000000"
+      "010000000000010001",
+      testMessage);
+}
+
+TEST_F(TcrMessageTest, testConstructorWithGinormousQueryEXECUTECQ_MSG_TYPE) {
```
Comment (echobravopapa) [inline src/cppcache/test/TcrMessage_unittest.cpp]: I'm not certain that this query's magnitude qualifies for ginormity... maybe we just call it test*64KiBQuery* - lest we set the bar low for ginormous...

File: src/cppcache/include/geode/DataOutput.hpp (hunk 1 above)
Comment (PivotalSarge) [inline src/cppcache/include/geode/DataOutput.hpp]: If the query string happened to be longer that 0xFFFF, it would be artificially truncated which often led to errors on the server side.

File: src/cppcache/test/DataOutputTest.cpp (hunk 2 above)
Comment (PivotalSarge) [inline src/cppcache/test/DataOutputTest.cpp]: getRemainingBufferLength() is the call that exhibited less-than-obvious behavior. In particular, it appears to be affected by some global state. Its return value was not predictable enough to be tested in that manner.

File: src/cppcache/test/TcrMessage_unittest.cpp (hunk 3 above)
Comment (PivotalSarge) [inline src/cppcache/test/TcrMessage_unittest.cpp]: I'm unaware of the SI definition of ginormous...

File: src/cppcache/include/geode/DataOutput.hpp [hunk 4]
```diff
@@ -378,9 +378,8 @@ class CPPCACHE_EXPORT DataOutput {
      */
   inline void writeFullUTF(const char* value, uint32_t length = 0) {
```
Comment (pivotal-jbarrett) [inline src/cppcache/include/geode/DataOutput.hpp]: If we are being sticklers and consistent this method has no business here. It is only used by the Message writeStringPart and should really be implemented there.

Comment (PivotalSarge) [inline src/cppcache/include/geode/DataOutput.hpp]: I agree and tried to move it to TcrMessage. However, its implementation depends on private members of DataOutput.

Comment (echobravopapa) [inline src/cppcache/include/geode/DataOutput.hpp]: ICKY, sounds like a separable refactoring chore...
